datafusion/physical-plan/src/aggregates/group_values/multi_group_by/primitive.rs

```diff
@@ -190,9 +191,13 @@ impl<T: ArrowPrimitiveType, const NULLABLE: bool> GroupColumn
             assert!(nulls.is_none(), "unexpected nulls in non nullable input");
         }
 
-        let arr = PrimitiveArray::<T>::new(ScalarBuffer::from(group_values), nulls);
+        let arr = PrimitiveArray::<T>::new(ScalarBuffer::from(group_values), nulls)
+            .with_data_type(data_type.clone());
+        let array_ref = Arc::new(arr) as ArrayRef;
+
         // Set timezone information for timestamp
-        Arc::new(arr.with_data_type(data_type))
+        adjust_output_array(&data_type, array_ref)
```

Comment:
If we need `adjust_output_array` then I think we don't need `with_data_type` to set the timezone information since it is handled  by `adjust_output_array` as well

Reply:
The precision and scale aren't kept in the generic when constructing the buffer, so i think i need t keep with_data_type

Reply:
It is kept in `data_type`. When you run `adjust_output_array`, the precision of Decimal should be set with `with_precision_and_scale`.

```rust
pub fn adjust_output_array(data_type: &DataType, array: ArrayRef) -> Result<ArrayRef> {
    let array = match data_type {
        DataType::Decimal128(p, s) => Arc::new(
            array
                .as_primitive::<Decimal128Type>()
                .clone()
                .with_precision_and_scale(*p, *s)?,
        ) as ArrayRef,
```

Reply:
Yeah that is what I thought as well but in https://github.com/apache/datafusion/pull/13564/commits/ea6f77aff4ae52faef6adf961697bb589820f844 it fails due to precision/scale errors due to the original array not having `.with_data_type` on it

Reply:
The error you mentioned can be fixed by adding support for Decimal256

Reply:
I tried backing out the use of adjust_data_type and I didn't see any failures locally 🤔